web-frontend for `social-app` with GraphQL types

Nirajsah opened the pull request:
## Motivation
This is a frontend react-app for the example social app. 

## Proposal
Changes:- 
`src/contract.rs` - Added functions to handle posts, likes and comments.

`src/lib.rs` - introduced structs for likes and comments, also updated previous structs to handle posts with images, likes and comments.

`src/state.rs` - updated states to handle updated `OwnPost` and `ReceivedPosts` objects. 

New:- 
`web-frontend` - includes necessary files to show a frontend for the social app. uses server generated GraphQL types. Updated `README.md` for better clarification.


## Test Plan
Test Case Update:- 
`tests/cross_chain.rs` - added imageUrl field to handle the test case.

`linera-core/src/unit_tests/./wasm_client_tests.rs` - added `image_url` field as `None` to handle the test case. 

All test cases passes successfully.

Twey commented on examples/social/web-frontend/public/manifest.json:
Let's get rid of this example manifest and the stock React icons :)

Nirajsah replied:
ok i'll do that

Nirajsah replied:
I have updated the test case. Test case failed in `linera-core/src/unit_tests/./wasm_client_tests.rs`.

Twey reviewed (commented):
This sounds really cool, thank you for the contribution! :grin: 

I tried to run it on my end to check it out, but I got a crypto error:

```
Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:79:19)
    at Object.createHash (node:crypto:139:10)
    at module.exports (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/util/createHash.js:90:53)
    at NormalModule._initBuildHash (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:401:16)
    at handleParseError (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:449:10)
    at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:481:5
    at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:342:12
    at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:221:10)
    at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:236:3
    at runSyncOrAsync (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:130:11)
    at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:232:2)
    at Array.<anonymous> (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
```

Do you know what that might be about?

I've left a few comments I spotted just scrolling through the code, but looking forward to being able to run it properly!  Our tests have some feedback as well :)

Nirajsah commented:
> This sounds really cool, thank you for the contribution! 😁
> 
> I tried to run it on my end to check it out, but I got a crypto error:
> 
> ```
> Error: error:0308010C:digital envelope routines::unsupported
>     at new Hash (node:internal/crypto/hash:79:19)
>     at Object.createHash (node:crypto:139:10)
>     at module.exports (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/util/createHash.js:90:53)
>     at NormalModule._initBuildHash (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:401:16)
>     at handleParseError (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:449:10)
>     at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:481:5
>     at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:342:12
>     at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:373:3
>     at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
>     at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:221:10)
>     at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:236:3
>     at runSyncOrAsync (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:130:11)
>     at iterateNormalLoaders (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:232:2)
>     at Array.<anonymous> (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
>     at Storage.finished (/home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
>     at /home/twey/dev/linera/protocol/nirajsah/social/examples/social/web-frontend/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
> ```
> 
> Do you know what that might be about?
> 
> I've left a few comments I spotted just scrolling through the code, but looking forward to being able to run it properly! Our tests have some feedback as well :)

are you getting this error when trying to run `npm install`

Nirajsah commented on examples/social/README.md:
```diff
@@ -125,42 +125,68 @@ Now make a post [in the 8080 tab](http://localhost:8080/chains/e476187f6ddfeb9d5
+            "index": 0
+          },
+          "text": "Linera Social is the new Mastodon!",
+          "imageUrl": null,
```
you're right, it shouldn't be null. its empty string. `""`

Nirajsah replied:
if we added `imageUrl` field when creating a post, it will show the url. otherwise it will show a empty string literal.

Nirajsah commented on examples/social/.DS_Store:
my bad. i didn't notice it. 🥲

Nirajsah commented:
i have noticed some of my mistakes! 

i'll make sure they won't happen again in future :)

Twey commented:
> are you getting this error when trying to run `npm install`?
No, it comes on `npm run start`.  The internet suggests it's a mismatch with my local OpenSSL, but I have quite a modern OpenSSL (3.0.14), which is recommended by Node as of a couple of years ago.  I think that one of the dependencies is relying on an old OpenSSL feature: as [this thread](https://stackoverflow.com/questions/70582072/npm-run-fails-with-err-ossl-evp-unsupported) suggests, setting `NODE_OPTIONS=--openssl-legacy-provider` lets me run it.  Now I get:

```
Failed to compile.

./src/GraphQLProvider.js
  Line 14:10:  'React' must be in scope when using JSX  react/react-in-jsx-scope
```

Twey commented on examples/social/.DS_Store:
Thank you for adding `.DS_Store` to our `.gitignore`!  This should make things easier for macOS users :)

This file is still here since it was added before the `.gitignore` change though.  Could we delete it from the PR (`git rm --cached .DS_Store`)?

Nirajsah replied:
Done :)

Nirajsah commented:
> > are you getting this error when trying to run `npm install`?
> 
> No, it comes on `npm run start`. The internet suggests it's a mismatch with my local OpenSSL, but I have quite a modern OpenSSL (3.0.14), which is recommended by Node as of a couple of years ago. I think that one of the dependencies is relying on an old OpenSSL feature: as [this thread](https://stackoverflow.com/questions/70582072/npm-run-fails-with-err-ossl-evp-unsupported) suggests, setting `NODE_OPTIONS=--openssl-legacy-provider` lets me run it. Now I get:
> 
> ```
> Failed to compile.
> 
> ./src/GraphQLProvider.js
>   Line 14:10:  'React' must be in scope when using JSX  react/react-in-jsx-scope
> ```

I have reproduced the issue. 
```Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:69:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/webpack/lib/util/createHash.js:90:53)
    at NormalModule._initBuildHash (/Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:401:16)
    at /Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:433:10
    at /Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/webpack/lib/NormalModule.js:308:13
    at /Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at /Users/nirajsah/Desktop/Pro/Rust/linera-protocol/examples/social/web-frontend/node_modules/babel-loader/lib/index.js:51:103
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.17.0
```

the issue is with `npm` to fix this issue you can run `npm audit fix --force` which solves this issue. i have tried it with other package managers and i don't see any issue there.

there shouldn't be any issues compiling. i have tried `npm run build` and deployed it locally. everything works fine. 
``` Failed to compile.

./src/GraphQLProvider.js
  Line 14:10:  'React' must be in scope when using JSX  react/react-in-jsx-scope
```

Twey commented:
With your help I got this working, and it looks pretty cool!

![social](https://github.com/user-attachments/assets/98ba56f0-bf62-4c4f-b094-e4cfd39add41)

The post button seems to work, but I don't see posts appear after I've posted them — do I have to do something special to refresh the list?

Twey commented on examples/social/web-frontend/README.md:
````diff
@@ -0,0 +1,70 @@
+To access the social app for a given Application Id, append the Application Id and port to the URI as so:
+
+```
+http://localhost:3000/7817752ff06b8266d77df8febf5c4b524cec096bd83dc54f989074fb94f833737ae984f32be2cee1dfab766fe2d0c726503c4d97117eb59023e9cc65a8ecd1f7000000000000000000000000000000007817752ff06b8266d77df8febf5c4b524cec096bd83dc54f989074fb94f833737ae984f32be2cee1dfab766fe2d0c726503c4d97117eb59023e9cc65a8ecd1f702000000000000000000000000000000&port=8000
````
This is not quite correct, the format is `https://localhost:8080/{chain_id}?app={app ID}&owner={owner ID}&port={node service port}`.  So running with the default settings from `examples/social/README.md` I had to navigate to:

```
http://localhost:8080/e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65?app=e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65010000000000000001000000e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65030000000000000000000000&owner=47f3d51e4315d75ceb5eafd05c63688cd150740adb8f26dcb8de3d54af07b992
```

Nirajsah replied:
:( i have updated the `README` file.

Nirajsah replied:
Also you can create a post with `imageUrl`, it looks good :)

Twey commented:
This looks pretty good to me.  We could do with updating the dependencies, but I think that's true of all our example frontends, so maybe better done in bulk.  Let's commit the results of `npm audit fix --force`, get the lints passing, and merge it! :)

Twey reviewed (changes requested):
Just a couple of minor fixes:
- `npm audit fix --force` seems to work to fix the dependencies, but we should commit the results
- the lints are currently failing

After that I think we're good to go :)

Nirajsah commented:
> This looks pretty good to me. We could do with updating the dependencies, but I think that's true of all our example frontends, so maybe better done in bulk in another PR.

I was thinking maybe we could migrate to `Bun.js` runtime instead of `node.js`. it's fast and doesn't have dependencies issues like `npm`.

Nirajsah commented:
> Just a couple of minor fixes:
> 
> * `npm audit fix --force` seems to work to fix the dependencies, but we should commit the results
> * the lints are currently failing
> 
> After that I think we're good to go :)

lints are failing showing `outdated README.md`. i have updated it, don't know where the issue is :)

Twey commented:
> I was thinking maybe we could migrate to `Bun.js` runtime instead of `node.js`. it's fast and doesn't have dependencies issues like `npm`.

I'm not against it, but I think that `npm` here is correctly highlighting an issue that Bun must be missing, which is that `react-scripts` (i.e. `create-react-app`) is not compatible with modern TypeScript because it's been dead as a project for a few years.  Ideally we should update all our frontends to not rely on `react-scripts` (maybe `react-scripts`' `eject` script is sufficient here, or at least a good starting point).

I've created a separate task (https://github.com/linera-io/linera-protocol/issues/2315) to track the migration from `react-scripts`, and if you'd like to do that in a separate PR I'd be happy to review :)  I'm also happy to swap to Bun if you want to PR that — I do like it better.  For this one, though, let's just commit the results of `npm audit fix --force` so that we can get it merged. 

> lints are failing showing outdated `README.md`. i have updated it, don't know where the issue is :)

Ah, I see the problem!

Our `README.md`s are actually autogenerated using a tool called [`cargo-rdme`](https://github.com/orium/cargo-rdme/tree/main) from the Rust documentation of the root module of the library (i.e. `src/lib.rs`).  Instead of editing `README.md` directly, you have to make your changes there, and then run `cargo rdme` to update the `README.md`.

Nirajsah commented:
> > I was thinking maybe we could migrate to `Bun.js` runtime instead of `node.js`. it's fast and doesn't have dependencies issues like `npm`.
> 
> I'm not against it, but I think that `npm` here is correctly highlighting an issue that Bun must be missing, which is that `react-scripts` (i.e. `create-react-app`) is not compatible with modern TypeScript because it's been dead as a project for a few years. Ideally we should update all our frontends to not rely on `react-scripts` (maybe `react-scripts`' `eject` script is sufficient here, or at least a good starting point).
> 
i use `vite+bun.js` so never faced any issues. 🥲 
> I've created a separate task (#2315) to track the migration from `react-scripts`, and if you'd like to do that in a separate PR I'd be happy to review :) I'm also happy to swap to Bun if you want to PR that — I do like it better. For this one, though, let's just commit the results of `npm audit fix --force` so that we can get it merged.
> 
yes we should update all the frontends. i would like to do that :). `fungible-app`s migration to `TS` is done. i'll fix if there is any dependencies issues and open a PR. should i continue with `react-scripts` or should i switch to `bun.js`. 

> > lints are failing showing outdated `README.md`. i have updated it, don't know where the issue is :)
> 
> Ah, I see the problem!
> 
> Our `README.md`s for Rust crates are actually autogenerated using a tool called [`cargo-rdme`](https://github.com/orium/cargo-rdme/tree/main) from the Rust documentation of the root module of the library (i.e. `src/lib.rs`). Instead of editing `README.md` directly, you have to make your changes there, and then run `cargo rdme` to update the `README.md`.
>
aah, i wasn't aware of that. that's really cool. i've updated it now.

Twey reviewed (changes requested):
We also have tests that check the commands in our READMEs are correctly executable, and it looks like your changes don't pass that test:

```
2024-07-31T11:27:13.0287407Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
2024-07-31T11:27:13.028504Z DEBUG linera_service::node_service: Request: Request { query: "mutation {\n    post(\n        text: \"Linera Social is the new Mastodon!\"\n        image: \"https://linera.org/img/logo.svg\" # optional\n    )\n}\n", operation_name: None, variables: Variables({}), extensions: {} }
2024-07-31T11:27:14.2291932Z 
2024-07-31T11:27:14.8839315Z 100   165    0     0  100   165      0    137  0:00:01  0:00:01 --:--:--   137
2024-07-31T11:27:14.8839830Z 100   285  100   120  100   165     64     88  0:00:01  0:00:01 --:--:--   153
2024-07-31T11:27:14.8840260Z 100   285  100   120  100   165     64     88  0:00:01  0:00:01 --:--:--   153
2024-07-31T11:27:14.8843286Z {"error":["Unknown argument \"image\" on field \"post\" of type \"OperationMutationRoot\". Did you mean \"imageUrl\"?"]}
2024-07-31T11:27:14.8849019Z null
2024-07-31T11:27:14.8854740Z + jobs -p
2024-07-31T11:27:14.8855937Z + xargs -r kill
2024-07-31T11:27:14.8922770Z + rm -rf /tmp/tmp.uunGnZBADG
2024-07-31T11:27:14.8946696Z thread 'test_script_in_readme_with_storage_service::social' panicked at linera-service/tests/readme_test.rs:51:5:
2024-07-31T11:27:14.8947598Z assertion failed: status.success()
```

Twey commented on examples/social/src/contract.rs:
```diff
@@ -80,9 +92,21 @@ impl Contract for SocialContract {
+    ) -> (Destination, Message) {
         let timestamp = self.runtime.system_time();
-        self.state.own_posts.push(OwnPost { timestamp, text });
+        let image = match image_url {
+            Some(image) => Some(image),
+            None => None,
+        };
```
Our lint checks are unhappy about this `match` because it's a no-op :)

Nirajsah replied:
the post expects an Option<String> type. but this match statement has no use because i am assigning variable image as an Option. if user does not provide a image_url then it should set it to `None`. 
but yeah having an option type will automatically set it to None if its not received. :)

Nirajsah commented:
> We also have tests that check the commands in our READMEs are correctly executable, and it looks like your changes don't pass that test:
> 
> ```
> 2024-07-31T11:27:13.0287407Z   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 2024-07-31T11:27:13.028504Z DEBUG linera_service::node_service: Request: Request { query: "mutation {\n    post(\n        text: \"Linera Social is the new Mastodon!\"\n        image: \"https://linera.org/img/logo.svg\" # optional\n    )\n}\n", operation_name: None, variables: Variables({}), extensions: {} }
> 2024-07-31T11:27:14.2291932Z 
> 2024-07-31T11:27:14.8839315Z 100   165    0     0  100   165      0    137  0:00:01  0:00:01 --:--:--   137
> 2024-07-31T11:27:14.8839830Z 100   285  100   120  100   165     64     88  0:00:01  0:00:01 --:--:--   153
> 2024-07-31T11:27:14.8840260Z 100   285  100   120  100   165     64     88  0:00:01  0:00:01 --:--:--   153
> 2024-07-31T11:27:14.8843286Z {"error":["Unknown argument \"image\" on field \"post\" of type \"OperationMutationRoot\". Did you mean \"imageUrl\"?"]}
> 2024-07-31T11:27:14.8849019Z null
> 2024-07-31T11:27:14.8854740Z + jobs -p
> 2024-07-31T11:27:14.8855937Z + xargs -r kill
> 2024-07-31T11:27:14.8922770Z + rm -rf /tmp/tmp.uunGnZBADG
> 2024-07-31T11:27:14.8946696Z thread 'test_script_in_readme_with_storage_service::social' panicked at linera-service/tests/readme_test.rs:51:5:
> 2024-07-31T11:27:14.8947598Z assertion failed: status.success()
> ```

I can't believe this :( 
Instead of `imageUrl` I wrote `image`

Twey commented:
Happens to the best of us!

You can run this test locally by running:
```shellsession
$ cargo test -p linera-service --test end_to_end_tests -- test_script_in_readme_with_storage_service::social
```

Twey commented:
Looks great, thank you! :grin:
